### PR TITLE
Rule based scanner

### PR DESCRIPTION
**Pull request by SUMUKHA-PK**
This PR when applied will be a complete implementation of a Rule Based Scanner.

Rule based scanner applies a set of rules one by one to figure out the matching rule for the set of runes, thus successfully scanning the input.

### CODE REVIEW

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 1]
```diff
@@ -15,13 +15,109 @@ func TestRuleBasedScanner(t *testing.T) {
...
+				token.New(1, 38, 37, 0, token.EOF, ""),
+			},
+		},
+		{
+			`SELECT FROM 'this \" can be anything'`,
+			ruleset.Default,
+			[]token.Token{
+				token.New(1, 1, 0, 6, token.KeywordSelect, "SELECT"),
+				token.New(1, 8, 7, 4, token.KeywordFrom, "FROM"),
+				token.New(1, 13, 12, 25, token.Literal, `'this \" can be anything'`),
```
[the remark this answers is not in the document]
Comment (TimSatke) [reply internal/parser/scanner/rule_based_scanner_test.go]: " basically has no special meaning in a ' string, which is why it doesn't need to be escaped

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 2]
```diff
@@ -15,13 +15,109 @@ func TestRuleBasedScanner(t *testing.T) {
 		want    []token.Token
 	}{
 		{
-			"SELECT FROM WHERE",
+			"SELECT FROM \"WHERE\"",
```
[the remark this answers is not in the document]
Comment (SUMUKHA-PK) [reply internal/parser/scanner/rule_based_scanner_test.go]: Sure

File: internal/parser/scanner/rule_based_scanner_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SUMUKHA-PK) [reply internal/parser/scanner/rule_based_scanner_test.go]: I was testing for correctness. Removed.

File: internal/parser/scanner/ruleset/ruleset_default.go [hunk 3]
```diff
@@ -86,7 +105,37 @@ func defaultDelimiterRule(s RuneScanner) (token.Type, bool) {
...
+				return token.Unknown, false
+			}
+			s.ConsumeRune()
+		} else if next == quoteType {
+			break
+		} else {
+			s.ConsumeRune()
+		}
+	}
+	s.ConsumeRune()
```
[the remark this answers is not in the document]
Comment (SUMUKHA-PK) [reply internal/parser/scanner/ruleset/ruleset_default.go]: Does it make a difference? Either ways it has to "return" right?

[the remark this answers is not in the document]
Comment (SUMUKHA-PK) [reply internal/parser/scanner/ruleset/ruleset_default.go]: Would it matter? It has to "return" anyway.

File: internal/parser/scanner/ruleset/ruleset_default.go [hunk 4]
```diff
@@ -62,18 +74,25 @@ func defaultKeywordsRule(s RuneScanner) (token.Type, bool) {
...
 	}
 	return token.Unknown, false
 }
 
-func defaultUnaryOperatorRule(s RuneScanner) (token.Type, bool) {
-	if next, ok := s.Lookahead(); ok && defaultUnaryOperator.Matches(next) {
+// defaultBinaryOperatorRule scans binary operators. Since binary opeators
+// can be of length one and two, it first checks for former, then latter
+// and allows the best of the two conditions.
+func defaultBinaryOperatorRule(s RuneScanner) (token.Type, bool) {
```
[the remark this answers is not in the document]
Comment (SUMUKHA-PK) [reply internal/parser/scanner/ruleset/ruleset_default.go]: Solved.

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 5]
```diff
@@ -30,7 +40,8 @@ func TestRuleBasedScanner(t *testing.T) {
 			[]token.Token{
 				token.New(1, 1, 0, 6, token.KeywordSelect, "SELECT"),
 				token.New(1, 8, 7, 4, token.KeywordFrom, "FROM"),
-				token.New(1, 13, 12, 6, token.Error, "unexpected token: '\"WHERE' at offset 12"),
+				token.New(1, 13, 12, 1, token.Error, `unexpected token: '"' at offset 12`),
```
Comment (TimSatke) [inline internal/parser/scanner/rule_based_scanner_test.go]: Why is there an unexpected token? `"WHERE"` should be a quoted `Literal`

Comment (TimSatke) [reply internal/parser/scanner/rule_based_scanner_test.go]: Sorry, GitHub view was confusing

**Pull request merged**